File: crypto-curriculum/modules/10-java/src/io/learncrypto/Hashes.java
package io.learncrypto;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.ByteArrayOutputStream;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;

/** SHA-256 and HKDF-SHA-256. RFC 5869 reference implementation. */
public final class Hashes {
    private Hashes() {}

    private static final byte[] EMPTY = new byte[0];

    public static byte[] sha256(byte[] data) throws GeneralSecurityException {
        return MessageDigest.getInstance("SHA-256").digest(data);
    }

    public static byte[] hkdfSha256(byte[] master, int outLen, byte[] salt, byte[] info)
            throws GeneralSecurityException {
        if (outLen <= 0) throw new IllegalArgumentException("outLen must be positive");
        if (outLen > 255 * 32) throw new IllegalArgumentException("outLen too large");
        byte[] effectiveSalt = salt != null ? salt : new byte[32];
        byte[] effectiveInfo = info != null ? info : EMPTY;
        // Extract
        byte[] prk = hmac(effectiveSalt, master);
        // Expand
        ByteArrayOutputStream okm = new ByteArrayOutputStream();
        byte[] prev = EMPTY;
        for (int counter = 1; okm.size() < outLen; counter++) {
            Mac m = Mac.getInstance("HmacSHA256");
            m.init(new SecretKeySpec(prk, "HmacSHA256"));
            m.update(prev);
            m.update(effectiveInfo);
            m.update((byte) counter);
            prev = m.doFinal();
            okm.write(prev, 0, Math.min(prev.length, outLen - okm.size()));
            if (counter > 255) throw new IllegalStateException("hkdf counter overflow");
        }
        return okm.toByteArray();
    }

    private static byte[] hmac(byte[] key, byte[] message) throws GeneralSecurityException {
        Mac m = Mac.getInstance("HmacSHA256");
        m.init(new SecretKeySpec(key, "HmacSHA256"));
        return m.doFinal(message);
    }
}

File: crypto-curriculum/modules/13-python/src/learncrypto/hashes.py
"""纯标准库实现的 SHA-256、HMAC-SHA-256、HKDF-SHA-256。"""

from __future__ import annotations

import hashlib
import hmac


def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def hmac_sha256(key: bytes, message: bytes) -> bytes:
    return hmac.new(key, message, hashlib.sha256).digest()


def hmac_sha256_verify(key: bytes, message: bytes, tag: bytes) -> bool:
    expected = hmac_sha256(key, message)
    return hmac.compare_digest(expected, tag)


def hkdf_sha256(master: bytes, out_len: int, salt: bytes = b"", info: bytes = b"") -> bytes:
    """RFC 5869 参考实现 HKDF-SHA-256。

    `salt=None` 会被替换为 32 个零字节字符串；`info` 是域分隔数据。
    """
    if out_len <= 0:
        raise ValueError("outLen must be positive")
    if out_len > 255 * 32:
        raise ValueError("outLen too large")

    if not salt:
        salt = bytes(32)

    prk = hmac.new(salt, master, hashlib.sha256).digest()
    out, prev, counter = bytearray(), b"", 1
    while len(out) < out_len:
        cur = hmac.new(prk, prev + info + bytes([counter]), hashlib.sha256).digest()
        out.extend(cur)
        prev = cur
        counter += 1
        if counter > 255:
            raise RuntimeError("HKDF overflow")
    return bytes(out[:out_len])
